src/sql/SqlResult.ts

```diff
@@ -0,0 +1,360 @@
+
+    /**
+     * Whether the result is closed or not. The result is closed if an update count or the last page is received.
+     * When true, there is no need to send the "cancel" request to the server.
+     */
+    private closed: boolean;
+
+    /**
+     * Row metadata of the result. Initially null.
+     */
+    private rowMetadata: SqlRowMetadata | null;
+
+    constructor(
+        private readonly sqlService: SqlServiceImpl,
+        private readonly serializationService: SerializationService,
+        private readonly connection: Connection,
+        private readonly queryId: SqlQueryId,
+        /** The page size used for pagination */
+        private readonly cursorBufferSize: number,
+        /** If true, SqlResult is an object iterable, otherwise SqlRow iterable */
+        private readonly returnRawResult: boolean,
+        private readonly clientUUID: UUID
+    ) {
+        this.closed = false;
+        this.last = false;
+        this.rowMetadata = null;
+        this.currentPage = null;
+        this.executeDeferred = deferredPromise<boolean>();
+        this.executeDeferred.promise.catch(() => {
+        });
+    }
+
+    /** This symbol is needed to be included to be an async iterable */
+    [Symbol.asyncIterator](): AsyncIterator<SqlRowType, SqlRowType, SqlRowType> {
+        const nextFn = this.next.bind(this);
+        return {
+            next: nextFn
+        }
+    }
+
+    /**
+     * Returns new result object. Useful for mocking. (Constructor mocking is hard/impossible)
+     */
+    static newResult(
+        sqlService: SqlServiceImpl,
+        serializationService: SerializationService,
+        connection: Connection,
+        queryId: SqlQueryId,
+        cursorBufferSize: number,
+        returnRawResult: boolean,
+        clientUUID: UUID
+    ) {
+        return new SqlResultImpl(sqlService, serializationService, connection, queryId, cursorBufferSize,
+            returnRawResult, clientUUID);
+    }
+
+    getUpdateCount(): Promise<Long> {
+        return this.executeDeferred.promise.then(() => {
+            return this.updateCount;
+        });
+    }
+
+    isRowSet(): Promise<boolean> {
+        return this.executeDeferred.promise.then(() => {
+            return this.updateCount === Long.fromInt(-1);
+        });
+    }
+
+    getRowMetadata(): Promise<SqlRowMetadata | null> {
+        return this.executeDeferred.promise.then(() => {
+            return this.rowMetadata;
+        });
+    }
+
+    close(): Promise<void> {
+        // Return the close promise if a close request is already started
+        if (this.closeDeferred?.promise) {
+            return this.closeDeferred.promise;
+        }
+
+        this.closeDeferred = deferredPromise<void>();
+
+        const error = new HazelcastSqlException(this.clientUUID, SqlErrorCode.CANCELLED_BY_USER, 'Cancelled by user');
+        // Reject execute with user cancellation error.
+        this.onExecuteError(error);
```

Comment:
OK about resolving/rejecting the promise twice (which I don't like, but it is in the spec so I can live with it :smile: ) but there are other side effects of this call, like setting `this.updateCount = Long.fromInt(-1);`. Can this have bad consequences?

Reply:
I don't think so. Just checked all the methods. calling close will cause all of them to reject, or no-op which ok I think

Reply:
One thing that comes to my mind. You get an update count for your query. You initiated close, then the result object will return -1 when you try to get update count. I am not sure if we are okay with this kind of behavior. What do you think?

Reply:
If user got an update count, the result should be already closed isn't it?  since:

```js
if (page.isLast()) {
            this.last = true;
            this.closed = true;
}
```

calling close again is no-op

Reply:
Ah sorry this is not called but this:
```js
if (rowMetadata !== null) { // Result that including rows
            this.rowMetadata = rowMetadata;
            this.onNextPage(rowPage);
            this.updateCount = Long.fromInt(-1);
        } else { // Result that including update count
            this.updateCount = updateCount;
            this.closed = true;
        }
```

else part is executed

Reply:
Hmm actually calling close() again is not no-op I think we can add that check to the close().

```js
if(this.closed){
    return
}
```

Reply:
I added returning a Promise.resolve() code, this way I think there will be no problem. similar check also exists on java side in close().